Msg Validation Rules for Boole Fork (#2)
ssvlabs/knowledge-base: p2p/MessageValidation/Rules.md

	ErrUnexpectedPrepareJustifications 	   = Error{text: "message has a prepare justification but it's not a proposal", reject: true}

	ErrPartialSigOneSigner              		        = Error{text: "partial signature message with len(signers) != 1", reject: true}

> 	ErrTooManyPartialSignatureMessages  		        = Error{text: "too many signatures for cluster in partial signature message"}

Reviewer: Any reason we're not setting `reject: true` on `ErrTooManyPartialSignatureMessages`?

The Duty Logic table classifies it as Reject, and the sibling `ErrTooManyEqualValidatorIndicesInPartialSignatures` rejects.

As it stands this is the only table/flag mismatch in the doc, and it makes sending oversized partial-sig messages penalty-free.

Reviewer 2: Node-side data point: the implementation already rejects here — [`ErrTooManySignaturesInPartialSigMessage`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/errors.go#L132) carries `reject: true`. So the Duty Logic table is the side the node agrees with; this error definition is the outlier.

	ErrPartialSigOneSigner              		= Error{text: "partial signature message with len(signers) != 1", reject: true}
	ErrTooManyPartialSignatureMessages  		= Error{text: "too many signatures for cluster in partial signature message"}

> 	ErrTooManyEqualValidatorIndexInPartialSignatures  = Error{text: "validator index appear 3 times in partial signature message", reject: true}

Reviewer 3: ```suggestion
	ErrTooManyEqualValidatorIndicesInPartialSignatures  = Error{text: "validator index appears too many times in partial signature message", reject: true}
```

Author: Agree, fixing it locally as it also appears in some other places, thanks!

The main structure is the `MessageValidation` structure which has a `ValidatePubsubMessage` function to serve as a handle for the GossipSub extended validator.

The function calls `ValidateMessage` which recursively calls every validation chain: Syntax -> Semantics -> QBFT Semantics | Partial Signature Semantics -> QBFT Logic -> Duty Rules.
All rules are verified against the peer-specific state, and after it:
- If there's any error, the appropriate `Ignore` or `Reject` is returned.
- Else, the message is again validated but this time against the global shared state.
This prevents the node from propagating duplicated messages. Then:
  - In case there's any error, `Ignore` is returned (even if the triggered rule has `reject: true`) as we can't blame the peer for the error.
  - Else, `Accept` is returned.

More about the global shared state validation is discussed [below](#global-shared-state-validations).

```go

const (

> 	MaxMsgSize                               = 9114816 // Source: https://github.com/ssvlabs/ssv-spec/blob/a65134ed45c932588c17d421173923f0216c6c73/types/spectest/tests/maxmsgsize/max_signed_ssv_message.go#L11

Reviewer: Could this lead to cheap memory pressure? 
`MaxMsgSize = 9114816` is only reachable by a proposal carrying two justifications, but it's applied to every message before the type-specific limits run.

A partial-sig or prepare message gets ~12.5x headroom to be buffered and SSZ-decoded before rejection, and this PR grew that by 84%. Should we consider applying the limit per `MsgType`, using the four constants that already exist in the spec repo?

Reviewer 2: Two more node deltas in this block, while it's being looked at:

1. **Classification**: `ErrSSVMessageDataTooBig` is Ignore here (no `reject` flag), but the node [rejects oversized `SSVMessage.Data`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/errors.go#L104) — and it already runs per-type gates on the inner data ([consensus](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/consensus_validation.go#L35-L40), [partial-sig](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/partial_validation.go#L33-L38)), though the outer envelope bound is still the single `MaxEncodedMsgSize`, so the buffering concern you raise applies to the node as-is. Reject is defensible (nothing honest exceeds a bound sitting above the spec max), but doc and node should pick the same side.
2. **Values**: the node doesn't enforce these spec constants directly — it [derives its own](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/const.go#L34-L69) with ~5% encoding overhead (740506 consensus / 763278 partial-sig / ~9.6 MB envelope vs 9114816 here), drift-guarded against the spec values in a test. So the numbers here are floors, not the enforced bounds.

```go

const (

> 	MaxMsgSize                               = 9114816

Reviewer 3: Can we add a comment with a link to how it can be calculated?

Author: Adding a reference to the spec test that computes it.

	// Check error
	if err != nil {
		return GetPubSubValidationResult(peerState, err)
	} else {
		// If the message is successful after all rules are tested on the peer-specific state,
		// test it as well on the global shared state to avoid propagating duplicated

> 		err = mv.ValidateAgainstGlobalSharedState(pmsg)

Reviewer: What happens if the global shared state is never written? 

The section below says it MUST be updated on acceptance of any message, but the pseudocode only updates the peer-specific state.
There's no `UpdateGlobalSharedState` call on the accept path, so as written the covert-attack check validates against a state that stays empty forever. 

Might also be worth defining what keys the global state uses in place of `peerID` (per-signer, presumably). If it's keyed only on duty/round/type, honest messages from different operators would collide and get silently ignored.

Reviewer 2: For reference, ssvlabs/ssv#2941 answers both questions in the implementation: the two-tier check runs inline per rule rather than as a second full pass. State is keyed msgID → signer → slot ([`SignerStateForSlotRound`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/signer_state.go#L12-L33)), each entry holding a per-peer map plus a `World` aggregate; a same-peer duplicate rejects, a cross-peer (`World`) hit ignores, and both tiers are written on the accept path ([consensus](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/consensus_validation.go#L365-L393), [partial-sig](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/partial_validation.go#L318-L346)).

So: per-signer keying, as you presumed — the doc could adopt that and add the missing global-state write on accept.

		if err != nil {
			return pubsub.ValidationIgnore
        }
		return pubsub.ValidationAccept
	}
}

> func GetPubSubValidatoinResult(peerState *PeerState, err error) pubsub.ValidationResult {

Reviewer 3: https://github.com/ssvlabs/knowledge-base/pull/3#discussion_r2769039654
```suggestion
func GetPubSubValidationResult(peerState *PeerState, err error) pubsub.ValidationResult {
```

Author: Fixed, thanks!

### Semantics General Rules

|     Verification     |           Error           | Classification | Explanation                                                                                                                                |
| -------------------- | ------------------------- | -------------- |--------------------------------------------------------------------------------------------------------------------------------------------|
| Signers in committee | ErrSignerNotInCommittee   | Reject         | Signers must belong to validator's or CommitteeID's committee.                                                                             |
| Different Domain     | ErrWrongDomain            | Ignore         | MsgID.Domain is different than self domain.                                                                                                |

> | Invalid Role         | ErrInvalidRole            | Reject         | MsgID.Role is wrong (not `RoleCommittee`, `RoleProposer`, `RoleValidatorRegistration`, `RoleVoluntaryExit`, or `RoleAggregatorCommittee`). |

Reviewer 3: Can we define that Alan roles are rejected during Boole and vice versa? (https://github.com/ssvlabs/ssv/pull/2503#issuecomment-3784620816)

Reviewer 2: It's kind of defined implicitly I think ?
- the updated rule set contains `RoleAggregatorCommittee` but not `RoleAggregator`/`RoleSyncCommitteeContribution`
- and although there is no explicit list of roles prior to this PR, if it were there - it would contain `RoleAggregator`/`RoleSyncCommitteeContribution` but not `RoleAggregatorCommittee`

But it would be nice for us devs to see "what changed" rather than trying to decipher the diff :) Maybe we could get that for the next fork ...

Reviewer 4: we have a taks to migrate to spec repo...
We need to think about how to make it as easy as possible.

Maybe you guys canuse this thread to write recommendations about what you like/don't like in the current structure

Reviewer 2: > Maybe you guys canuse this thread to write recommendations about what you like/don't like in the current structure

Not strictly about the structure of these files, but I would think some sort of writeup in English about every rule added/changed/deleted (**before** vs **after** the fork) would be useful to have whenever  we are doing these changes

| Invalid Role         | ErrInvalidRole            | Reject         | MsgID.Role is wrong (not `RoleCommittee`, `RoleProposer`, `RoleValidatorRegistration`, `RoleVoluntaryExit`, or `RoleAggregatorCommittee`). |
| Validator exists     | ErrUnknownValidator       | Ignore         | If MsgID.SenderID is a validator, it must exist.                                                                                           |
| Active Validator ID  | ErrValidatorNotAttesting  | Ignore         | If MsgID.SenderID is a validator, it must be active active validator.                                                                      |
| Validator Liquidated | ErrValidatorLiquidated    | Ignore         | If MsgID.SenderID is a validator, it must not be liquidated.                                                                               |
| CommitteeID exists   | ErrNonExistentCommitteeID | Ignore         | If MsgID.SenderID is a committee, it must exist.                                                                                           |

> | Wrong topic          | ErrIncorrectTopic         | Ignore         | The message should be sent in the correct topic                                                                                            |

Reviewer: Might be worth writing the topic rule down here, the PR description defines `/ssv/<ethereum_network_name>/boole/<subnet>`, but neither the format nor the MsgID-to-subnet derivation appears anywhere in the doc, so `CorrectTopic` isn't implementable from this spec alone.

 It also decides whether the subnet derives from the full 48-byte SenderID or the trailing 32-byte CommitteeID, which interacts with the encoding question above.

	}

> 	senderID := signedSSVMessage.SSVMessage.MsgID.GetSenderID()

Reviewer: Do we need to spell out the SenderID encoding here? 

The PR description says the CommitteeID must be encoded with a 16-byte 0x00 prefix, but that rule doesn't appear anywhere in the doc — only the "unknown committee → ignore" half made it in. 

Since every stateful limit is keyed on `MsgID`, leaving the prefix bytes unconstrained would let a committee operator mint 2^128 aliases of the same committee, each with a fresh rate-limit bucket; and if implementations instead compare all 48 bytes, differently-padded honest messages get dropped. 

A rule like "for committee roles, `SenderID[0:16]` must be all-zero (Reject otherwise) and the CommitteeID is `SenderID[16:48]`" would pin it down.

| ------------------------- | ----------------------------- | -------------- |------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Already advanced slot     | ErrSlotAlreadyAdvanced        | Ignore         | (Non-committee roles) Signer already advanced to later slot.                                                                                                                                                     |
| Invalid role for consensus  | ErrUnexpectedConsensusMessage | Reject         | SSVMessage.MsgID.Role must not be ValidatorRegistration or VoluntaryExit.                                                                                                                                        |
| No beacon duty            | ErrNoDuty                     | Ignore         | If Proposal duty, check if duty exists with beacon node.                                                                                                                                                         |

> | Slot not in time for role | ErrEarlySlotMessage or ErrLateSlotMessage              | Ignore         | Current time must be between duty's starting time and<br> +34 (committee and aggregator committee) or +3 (else) slots.                                                                                           |

Reviewer: Worth considering whether the merged role should inherit all the committee limits? 

Compared to Alan's `RoleAggregator`/`RoleSyncCommitteeContribution`, `RoleAggregatorCommittee` moves from a +3 to a +34 slot window, round cut-off 6 to 12, duties per epoch 2 to 2*V, and the `HasSyncCommitteeDuty` existence check was removed with no replacement. 

Sync-committee membership is known a period ahead and attester duties an epoch ahead, so some duty-existence check still seems feasible, or at least a note on why the committee windows are safe for this traffic.

| Too many duties per epoch | ErrTooManyDutiesPerEpoch      | Ignore         | If role is either voluntary exit and validator registration,<br> it's allowed 2 duties per epoch. Else if committee or aggregator committee,<br> 2*V (if no validator is doing sync committee).<br> Else accept. |

> | Valid round for role      | ErrRoundTooHigh               | Reject         | For committee and aggregator committee, round can go up to 12. Else, it can go up to 6.                                                                                                                          |

Reviewer 2: The node's round limits diverge from this row in three ways ([`maxRound`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/consensus_validation.go#L427-L438), [`roundBelongsToAllowedSpread`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/consensus_validation.go#L531-L569)):

1. **Proposer cut-off is 2, not 6** — presumably because a block that missed its slot can't be published anyway (the in-code TODO says the values are still under review).
2. **`ErrRoundTooHigh` is Ignore in the node**, not Reject as classified here — deliberate softening while the limits are in doubt (the TODO notes stage traffic already exceeded the aggregator limit).
3. **The estimated-round spread is [-2, +1]** ([`allowedRoundsInPast` / `allowedRoundsInFuture`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/const.go#L14-L15)), not the "spread of 1 round" the QBFT-logic table describes — and the node skips the spread check for Proposer entirely (ssvlabs/ssv#2429).

If the doc is normative, the node has three deltas to fix; if descriptive, these want recording — either way the two shouldn't disagree silently.

		return ErrUnexpectedConsensusMessage
	}

    // Rule: For proposer duty, we check if the validator is assigned to it
    if err := mv.ValidProposerDuty(signedSSVMessage.SSVMessage.MsgID.GetSenderID(), signedSSVMessage.SSVMessage.MsgID.GetRoleType(), phase0.Slot(qbftMessage.Height)); err != nil {
		return err
    }

	// Rule: current slot(height) must be between duty's starting slot and:
	// - duty's starting slot + 34 (committee and aggregator committee)
	// - duty's starting slot + 3 (other types)

> 	if err != mv.ValidDutySlot(peerID, phase0.Slot(qbftMessage.Height), signedSSVMessage.SSVMessage.MsgID.GetRoleType()); err != nil {

Reviewer 2: Pseudo-code nits, collected in one thread:

- Here: comparison instead of assignment —
```suggestion
	if err := mv.ValidDutySlot(peerID, phase0.Slot(qbftMessage.Height), signedSSVMessage.SSVMessage.MsgID.GetRoleType()); err != nil {
```
- L717: `consensusMessage.Round` — undefined in this listing; the local is `qbftMessage`.
- L843-853: `ValidProposerDuty()` takes no parameters but uses `signedSSVMessage`/`qbftMessage`, while both call sites pass three arguments.
- L252-253 (syntax table): "must be >= 0" for signers/signatures should be "> 0" (the code errors on `== 0`).

| No PartialSignatureMessage | ErrNoPartialSignatureMessages | Reject | Message must have at least one PartialSignatureMessage.                                                                                                                                                                                                                                                                             |
| Wrong BLS Signature Size   | ErrWrongBLSSignatureSize      | Reject | $\forall i$ PartialSignatureMessages.Message[i].Signature must have the correct length.                                                                                                                                                                                                                                             |
| Inconsistent signer        | ErrInconsistentSigners        | Reject | $\forall i$ PartialSignatureMessages.Message[i].Signer must be the same as the<br> SignedSSVMessage.OperatorIDs[i].                                                                                                                                                                                                                 |

> | Validator's index mismatch | ErrValidatorIndexMismatch     | Ignore | $\forall i$ PartialSignatureMessages.Message[i].ValidatorIndex must belong to SSVMessage.SenderID().                                                                                                                                                                                                                                |

Reviewer 2: The node deliberately doesn't implement this for committee roles: [the check is skipped](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/partial_validation.go#L144-L154) with the rationale that operators aren't assumed synced on each other's validator sets, so asserting `ValidatorIndex ∈ CommitteeID's validators` could drop honest messages during registry lag; it's enforced for non-committee roles only. The volume rows below still bound the damage, but the indices themselves go unchecked for committee duties.

Worth a footnote recording the relaxation — or, if this row is the target behavior, a node-side issue to add the check.

Reviewer: Following up from ssvlabs/ssv#2941 ([thread](https://github.com/ssvlabs/ssv/pull/2941#discussion_r3665814973)): the decision there was to keep the relaxation.

The in-code rationale (operators aren't assumed synced on each other's validator sets) still holds, and the per-index count limits bound the exposure. So the doc should own the record: maybe a note on this row that for committee and aggregator-committee roles the check is deliberately skipped, and Ignore applies only to the non-committee roles where it runs.

| Too many duties per epoch    | ErrTooManyDutiesPerEpoch | Ignore         | If role is either aggregator, voluntary exit and validator registration,<br> it's allowed 2 duties per epoch. Else if committee or aggregator committee,<br> 2*V (if no validator is doing sync committee).<br> Else accept.                                                                                                                          |
| Too many partial signatures  | ErrTooManyPartialSignatureMessages                | Reject         | For the committee role, it's allowed $min(2*V, V + $ SYNC_COMMITTEE_SIZE $)$ <br> where $V$ is the number of committee's validators.<br> For the aggregator committee role, it's allowed $min((1+4)*V, V + 4\times$ SYNC_COMMITTEE_SIZE $)$ <br> where $V$ is the number of committee's validators.<br> Else, only 1.                                 |

> | Triple validator index       | ErrTooManyEqualValidatorIndexInPartialSignatures                | Reject         | For the committee (resp. aggregator committee) role, a validator index can not be associated to more than 2 (resp. 5) signatures.                                                                                                                                                                                                                     |

Reviewer 3: ```suggestion
| Too many equal validator indices       | ErrTooManyEqualValidatorIndexInPartialSignatures | Reject         | A validator index can not be associated with more than 2 signatures for the committee role and more than 5 for the aggregator committee role.                                                                                                                                                                                                                     |
```

| Slot not in time for role        | ErrEarlySlotMessage or ErrLateSlotMessage          | Ignore         | Current time must be between duty's starting time and<br> +34 (committee and aggregator committee) or +3 (else) slots.                                                                                                                                                                                                                              |
| Too many duties per epoch        | ErrTooManyDutiesPerEpoch                           | Ignore         | If role is either aggregator, voluntary exit and validator registration,<br> it's allowed 2 duties per epoch. Else if committee or aggregator committee,<br> 2*V (if no validator is doing sync committee).<br> Else accept.                                                                                                                        |

> | Too many partial signatures      | ErrTooManyPartialSignatureMessages                 | Reject         | For the committee role, it's allowed $min(2*V, V + $ SYNC_COMMITTEE_SIZE $)$ <br> where $V$ is the number of committee's validators.<br> For the aggregator committee role, it's allowed $min((1+4)*V, V + 4 \times$ SYNC_COMMITTEE_SIZE $)$ <br> where $V$ is the number of committee's validators.<br> Else, only 1.                              |

Reviewer 4: why this change @MatheusFranco99 ?

Author: Note that it's a repetition (already had it in the previous sentence).

| No beacon duty                   | ErrNoDuty                                          | Ignore         | If Proposal duty, check if duty exists with beacon node.                                                                                                                                                                                                                                                                                            |
| Invalid signature type count     | ErrInvalidPartialSignatureTypeCount                | Reject         | It allows only:<br> 1 PostConsensusPartialSig, for Committee duty,<br> 1 RandaoPartialSig and 1 PostConsensusPartialSig for Proposer,<br> 1 AggregatorCommitteePartialSig and 1 PostConsensusPartialSig for AggregatorCommittee,<br> 1 ValidatorRegistrationPartialSig for Validator Registration,<br> 1 VoluntaryExitPartialSig for Voluntary Exit.|

> | Slot not in time for role        | ErrEarlySlotMessage or ErrLateSlotMessage          | Ignore         | Current time must be between duty's starting time and<br> +34 (committee and aggregator committee) or +3 (else) slots.                                                                                                                                                                                                                              |

Reviewer 2: The node doesn't enforce "+3 (else)" for ValidatorRegistration / VoluntaryExit: [`messageLateness`](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/common_checks.go#L40-L55) has no case for them and returns 0 — they're bounded only by the early check, the slot-advance rule, and the duty caps (a legacy exemption; registrations aren't slot-anchored the way duties are). If +3 is meant to be normative for these roles it's a node change; otherwise this row (and the `ErrSlotNotInTime`/`ErrLateSlotMessage` texts) should carve them out.

Reviewer: Closing the loop from ssvlabs/ssv#2941 ([thread](https://github.com/ssvlabs/ssv/pull/2941#discussion_r3665814911)): we decided to keep the exemption deliberate rather than bound VR/VE.

The node now records it explicitly (own case in `messageLateness` + corrected `ErrLateSlotMessage` text, ssvlabs/ssv@207790071). So this row should record it too, something like:
 "+34 (committee and aggregator committee) or +3 (proposer and sync committee contribution) slots; ValidatorRegistration and VoluntaryExit have no late bound. Only the early check and per-epoch duty limits apply."

> | Too many duties per epoch        | ErrTooManyDutiesPerEpoch                           | Ignore         | If role is either aggregator, voluntary exit and validator registration,<br> it's allowed 2 duties per epoch. Else if committee or aggregator committee,<br> 2*V (if no validator is doing sync committee).<br> Else accept.                                                                                                                        |

Reviewer: Should this still say "aggregator"? 

The QBFT counterpart of this row dropped it (there's no `RoleAggregator` in Boole), and as written "aggregator" effectively lands in both branches of the rule. The pseudocode comment also allows 2 only for voluntary exit and validator registration.

> | Too many duties per epoch        | ErrTooManyDutiesPerEpoch                           | Ignore         | If role is either aggregator, voluntary exit and validator registration,<br> it's allowed 2 duties per epoch. Else if committee or aggregator committee,<br> 2*V (if no validator is doing sync committee).<br> Else accept.                                                                                                                        |

Reviewer 2: Adjacent delta: the node's VoluntaryExit limit isn't a flat 2 — it's [the number of exit requests actually observed for that pubkey](https://github.com/ssvlabs/ssv/blob/33f8057b6ef5c8957aef569a658452de5bdf610e/message/validation/common_checks.go#L91-L100) via the duty store (exits are event-driven, so the node knows how many are legitimate). The flat 2 matches the node only for ValidatorRegistration.

> | Too many partial signatures      | ErrTooManyPartialSignatureMessages                 | Reject         | For the committee role, it's allowed $min(2*V, V + $ SYNC_COMMITTEE_SIZE $)$ <br> where $V$ is the number of committee's validators.<br> For the aggregator committee role, it's allowed $min((1+4)*V, V + 4 \times$ SYNC_COMMITTEE_SIZE $)$.<br> Else, only 1.                              |

Reviewer: Should we consider tightening the aggregator committee bound to $min(5V, V + $ SYNC_COMMITTEE_SIZE $)$? 

There are only 512 sync-committee positions network-wide and each position maps to exactly one subnet, so the sum of subnets across all validators is bounded by SYNC_COMMITTEE_SIZE, not 4x it.
The $4 \times$ term double-counts. 

The committee-role formula in this same row is derived the tight way, so the two are inconsistent. It also matters downstream: the loose bound is what inflates the max partial-sig payload to ~727 KB, vs ~506 KB with the tight derivation.

			}
			type SSVMessage struct {
				MsgType MsgType
				MsgID   MessageID -> Must be assigned to duty if role is Proposal
				Data 	[]byte
			}
			type PartialSignatureMessages struct {
				Type     PartialSigMsgType -> Message count rules
				Slot     phase0.Slot -> Must belong to allowed spread, Satisfies a maximum number of duties per epoch for role, Must not be "old"

> 				Messages []*PartialSignatureMessage -> Valid number of signatures (3 cases: committee duty, aggregator committee duty, others)

Reviewer 4: I find what is written in parentehsis confusing?

Author: Agree. I meant there are 3 cases for validating the number of signatures in a msg (one for committee duty, one for agg comm duty, and another for other duties).
Tried to make it clearer:
```go
     // ...
     Messages []*PartialSignatureMessage // -> Valid number of signatures
     // (3 cases: min(2*V, V + SYNC_COMMITTEE_SIZE) for committee duty;
     // min((1+4)*V, V + 4 * SYNC_COMMITTEE_SIZE) for aggregator committee duty;
     // 1 for other duty types)
     // ...
```
Thanks!
